Periodic kernel for multivariate input

# 🐛 Bug

Implementation of PeriodicKernel is not positive definite in case of multivariate input
Here is how I checked it:

** Code snippet to reproduce **
```python
import torch
from gpytorch import kernels

X = torch.randn(1000, 2, dtype=torch.float64)
kernel = kernels.PeriodicKernel().double()
with torch.no_grad():
    K = kernel(X, X).evaluate()
    eig = torch.symeig(K)[0]

    negative_eig = eig[eig < 0]
    assert torch.allclose(negative_eig, torch.zeros_like(negative_eig))
```

** Stack trace/error message **
```
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
<ipython-input-2-1d738d7c73d9> in <module>
      9 
     10     negative_eig = eig[eig < 0]
---> 11     assert torch.allclose(negative_eig, torch.zeros_like(negative_eig))

AssertionError: 
```

## Expected Behavior

Positive definite kernel must have non-negative eigenvalues.

##
The implementation is correct for one-dimensional input. For multidimensional input it should be product of one-dimensional kernels, however, it calculates the norm first and applies sine and exponent to the norm which is not correct: see eq. (47) on page 15 of
> MacKay, D. J. (1998). Introduction to Gaussian processes. NATO ASI Series F Computer and Systems Sciences, 168, 133-166.





## System information

**Please complete the following information:**
- gpytorch=0.3.4
- torch=1.1.0
- Ubuntu 19.04

--- a/gpytorch/kernels/periodic_kernel.py
+++ b/gpytorch/kernels/periodic_kernel.py
@@ -16,9 +16,8 @@ class PeriodicKernel(Kernel):
 
         \begin{equation*}
             k_{\text{Periodic}}(\mathbf{x_1}, \mathbf{x_2}) = \exp \left(
-            -\frac{1}{2} \sum_i \left(
-            \frac{\sin \left( \frac{\pi}{\lambda} (\mathbf{x_{1,i}} - \mathbf{x_{2,i}}\right)}{p}
-            \right)^2
+            -2 \sum_i
+            \frac{\sin ^2 \left( \frac{\pi}{p} (\mathbf{x_{1,i}} - \mathbf{x_{2,i}} ) \right)}{\lambda}
             \right)
         \end{equation*}
 
@@ -27,9 +26,10 @@ class PeriodicKernel(Kernel):
     * :math:`p` is the period length parameter.
     * :math:`\lambda` is a lengthscale parameter.
 
-    Equation is taken from [David Mackay's Introduction to Gaussian Processes equation 47]
+    Equation is based on [David Mackay's Introduction to Gaussian Processes equation 47]
     (http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.81.1927&rep=rep1&type=pdf)
-    albeit without feature-specific lengthscales and period lengths.
+    albeit without feature-specific lengthscales and period lengths. The exponential
+    coefficient was changed and lengthscale is not squared to maintain backwards compatibility
 
     .. note::
 
@@ -120,7 +120,7 @@ class PeriodicKernel(Kernel):
         else:
             raise ValueError("x.dims() must be 2 or 3")
         diff = diff * math.pi / self.period_length.unsqueeze(-1)
-        res = diff.sin().div(self.lengthscale.unsqueeze(-1)).pow(2).sum(dim=-1).mul(-0.5).exp_()
+        res = diff.sin().pow(2).sum(dim=-1).div(self.lengthscale).mul(-2.0).exp_()
         if diag:
             res = res.squeeze(0)
         return res

--- a/gpytorch/kernels/periodic_kernel.py
+++ b/gpytorch/kernels/periodic_kernel.py
@@ -107,10 +107,18 @@ class PeriodicKernel(Kernel):
         self.initialize(raw_period_length=self.raw_period_length_constraint.inverse_transform(value))
 
     def forward(self, x1, x2, diag=False, **params):
-        x1_ = x1.div(self.period_length)
-        x2_ = x2.div(self.period_length)
-        diff = self.covar_dist(x1_, x2_, diag=diag, **params)
-        res = torch.sin(diff.mul(math.pi)).pow(2).mul(-2 / self.lengthscale).exp_()
+        # x1_ = x1.div(self.period_length)
+        # x2_ = x2.div(self.period_length)
+        # diff = self.covar_dist(x1_, x2_, diag=diag, **params)
+        if x1.dim() == 3 and x2.dim() == 3:  # batched operation
+            diff = x1.unsqueeze(2) - x2.unsqueeze(1)
+        elif x1.dim() == 2 and x2.dim() == 2:
+            diff = x1.unsqueeze(1) - x2.unsqueeze(0)
+        else:
+            raise ValueError("x.dims() must be 2 or 3")
+        diff = diff * math.pi / self.period_length
+        res = diff.sin().div(self.lengthscale).pow(2).sum(dim=-1).mul(-0.5).exp_()
+        # res = torch.sin(diff.mul(math.pi)).pow(2).mul(-2 / self.lengthscale).exp_()
         if diag:
             res = res.squeeze(0)
         return res

--- a/gpytorch/kernels/periodic_kernel.py
+++ b/gpytorch/kernels/periodic_kernel.py
@@ -14,16 +14,22 @@ class PeriodicKernel(Kernel):
 
     .. math::
 
-       \begin{equation*}
-          k_{\text{Periodic}}(\mathbf{x_1}, \mathbf{x_2}) = \exp \left(
-            \frac{2 \sin^2 \left( \pi \Vert \mathbf{x_1} - \mathbf{x_2} \Vert_1 / p \right) }
-            { \ell^2 } \right)
-       \end{equation*}
+    \begin{equation*}
+        k_{\text{Periodic}}(\mathbf{x_1}, \mathbf{x_2}) = \exp \left(
+        -\frac{1}{2} \sum_i \left(
+        \frac{\sin \left( \frac{\pi}{\lambda} (\mathbf{x_{1,i}} - \mathbf{x_{2,i}}\right)}{p}
+        \right)^2
+        \right)
+    \end{equation*}
 
     where
 
-    * :math:`p` is the periord length parameter.
-    * :math:`\ell` is a lengthscale parameter.
+    * :math:`p` is the period length parameter.
+    * :math:`\lambda` is a lengthscale parameter.
+
+    Equation is taken from [David Mackay's Introduction to Gaussian Processes equation 47]
+    (http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.81.1927&rep=rep1&type=pdf)
+    albeit without feature-specific lengthscales and period lengths.
 
     .. note::
 
@@ -32,7 +38,7 @@ class PeriodicKernel(Kernel):
 
     .. note::
 
-        This kernel does not have an ARD lengthscale option.
+        This kernel does not have an ARD lengthscale or period length option.
 
     Args:
         :attr:`batch_shape` (torch.Size, optional):

--- a/gpytorch/kernels/periodic_kernel.py
+++ b/gpytorch/kernels/periodic_kernel.py
@@ -14,13 +14,13 @@ class PeriodicKernel(Kernel):
 
     .. math::
 
-    \begin{equation*}
-        k_{\text{Periodic}}(\mathbf{x_1}, \mathbf{x_2}) = \exp \left(
-        -\frac{1}{2} \sum_i \left(
-        \frac{\sin \left( \frac{\pi}{\lambda} (\mathbf{x_{1,i}} - \mathbf{x_{2,i}}\right)}{p}
-        \right)^2
-        \right)
-    \end{equation*}
+        \begin{equation*}
+            k_{\text{Periodic}}(\mathbf{x_1}, \mathbf{x_2}) = \exp \left(
+            -\frac{1}{2} \sum_i \left(
+            \frac{\sin \left( \frac{\pi}{\lambda} (\mathbf{x_{1,i}} - \mathbf{x_{2,i}}\right)}{p}
+            \right)^2
+            \right)
+        \end{equation*}
 
     where
 

--- a/gpytorch/kernels/periodic_kernel.py
+++ b/gpytorch/kernels/periodic_kernel.py
@@ -113,13 +113,9 @@ class PeriodicKernel(Kernel):
         self.initialize(raw_period_length=self.raw_period_length_constraint.inverse_transform(value))
 
     def forward(self, x1, x2, diag=False, **params):
-        if x1.dim() == 3 and x2.dim() == 3:  # batched operation
-            diff = x1.unsqueeze(2) - x2.unsqueeze(1)
-        elif x1.dim() == 2 and x2.dim() == 2:
-            diff = x1.unsqueeze(1) - x2.unsqueeze(0)
-        else:
-            raise ValueError("x.dims() must be 2 or 3")
-        diff = diff * math.pi / self.period_length.unsqueeze(-1)
+        x1_ = x1.div(self.period_length).mul(math.pi)
+        x2_ = x2.div(self.period_length).mul(math.pi)
+        diff = x1_.unsqueeze(-2) - x2_.unsqueeze(-3)
         res = diff.sin().pow(2).sum(dim=-1).div(self.lengthscale).mul(-2.0).exp_()
         if diag:
             res = res.squeeze(0)

--- a/gpytorch/kernels/periodic_kernel.py
+++ b/gpytorch/kernels/periodic_kernel.py
@@ -107,18 +107,14 @@ class PeriodicKernel(Kernel):
         self.initialize(raw_period_length=self.raw_period_length_constraint.inverse_transform(value))
 
     def forward(self, x1, x2, diag=False, **params):
-        # x1_ = x1.div(self.period_length)
-        # x2_ = x2.div(self.period_length)
-        # diff = self.covar_dist(x1_, x2_, diag=diag, **params)
         if x1.dim() == 3 and x2.dim() == 3:  # batched operation
             diff = x1.unsqueeze(2) - x2.unsqueeze(1)
         elif x1.dim() == 2 and x2.dim() == 2:
             diff = x1.unsqueeze(1) - x2.unsqueeze(0)
         else:
             raise ValueError("x.dims() must be 2 or 3")
-        diff = diff * math.pi / self.period_length
-        res = diff.sin().div(self.lengthscale).pow(2).sum(dim=-1).mul(-0.5).exp_()
-        # res = torch.sin(diff.mul(math.pi)).pow(2).mul(-2 / self.lengthscale).exp_()
+        diff = diff * math.pi / self.period_length.unsqueeze(-1)
+        res = diff.sin().div(self.lengthscale.unsqueeze(-1)).pow(2).sum(dim=-1).mul(-0.5).exp_()
         if diag:
             res = res.squeeze(0)
         return res

--- a/test/kernels/test_periodic_kernel.py
+++ b/test/kernels/test_periodic_kernel.py
@@ -20,8 +20,8 @@ class TestPeriodicKernel(unittest.TestCase):
         actual = torch.zeros(3, 2)
         for i in range(3):
             for j in range(2):
-                val = (torch.sin(math.pi / period * (a[i] - b[j])) / lengthscale) ** 2
-                actual[i, j] = torch.exp(-0.5 * val).item()
+                val = 2 * torch.pow(torch.sin(math.pi * (a[i] - b[j]) / 3), 2) / lengthscale
+                actual[i, j] = torch.exp(-val).item()
 
         res = kernel(a, b).evaluate()
         self.assertLess(torch.norm(res - actual), 1e-5)
@@ -72,7 +72,7 @@ class TestPeriodicKernel(unittest.TestCase):
         for k in range(2):
             diff = a[k].unsqueeze(1) - b[k].unsqueeze(0)
             diff = diff * math.pi / period[k].unsqueeze(-1)
-            actual[k] = diff.sin().div(lengthscale[k]).pow(2).sum(dim=-1).mul(-0.5).exp_()
+            actual[k] = diff.sin().pow(2).sum(dim=-1).div(lengthscale[k]).mul(-2.0).exp_()
 
         res = kernel(a, b).evaluate()
         self.assertLess(torch.norm(res - actual), 1e-5)

--- a/test/kernels/test_periodic_kernel.py
+++ b/test/kernels/test_periodic_kernel.py
@@ -60,20 +60,22 @@ class TestPeriodicKernel(unittest.TestCase):
         res = kernel(a, b).evaluate()
         self.assertLess(torch.norm(res - actual), 1e-5)
 
-    # def test_batch_separate(self):
-    #     a = torch.tensor([[4, 2, 8], [1, 2, 3]], dtype=torch.float).view(2, 3, 1)
-    #     b = torch.tensor([[0, 2], [-1, 2]], dtype=torch.float).view(2, 2, 1)
-    #     period = torch.tensor([1, 2], dtype=torch.float).view(2, 1, 1)
-    #     lengthscale = torch.tensor([2, 1], dtype=torch.float).view(2, 1, 1)
-    #     kernel = PeriodicKernel(batch_shape=torch.Size([2])).initialize(lengthscale=lengthscale, period_length=period)
-    #     kernel.eval()
-    #
-    #     actual = torch.zeros(2, 3, 2)
-    #     for k in range(2):
-    #         actual[k] = kernel(a[k], b[k]).evaluate()
-    #
-    #     res = kernel(a, b).evaluate()
-    #     self.assertLess(torch.norm(res - actual), 1e-5)
+    def test_batch_separate(self):
+        a = torch.tensor([[4, 2, 8], [1, 2, 3]], dtype=torch.float).view(2, 3, 1)
+        b = torch.tensor([[0, 2], [-1, 2]], dtype=torch.float).view(2, 2, 1)
+        period = torch.tensor([1, 2], dtype=torch.float).view(2, 1, 1)
+        lengthscale = torch.tensor([2, 1], dtype=torch.float).view(2, 1, 1)
+        kernel = PeriodicKernel(batch_shape=torch.Size([2])).initialize(lengthscale=lengthscale, period_length=period)
+        kernel.eval()
+
+        actual = torch.zeros(2, 3, 2)
+        for k in range(2):
+            diff = a[k].unsqueeze(1) - b[k].unsqueeze(0)
+            diff = diff * math.pi / period[k].unsqueeze(-1)
+            actual[k] = diff.sin().div(lengthscale[k]).pow(2).sum(dim=-1).mul(-0.5).exp_()
+
+        res = kernel(a, b).evaluate()
+        self.assertLess(torch.norm(res - actual), 1e-5)
 
 
 if __name__ == "__main__":

--- a/test/kernels/test_periodic_kernel.py
+++ b/test/kernels/test_periodic_kernel.py
@@ -20,11 +20,30 @@ class TestPeriodicKernel(unittest.TestCase):
         actual = torch.zeros(3, 2)
         for i in range(3):
             for j in range(2):
-                val = 2 * torch.pow(torch.sin(math.pi * (a[i] - b[j]) / 3), 2) / lengthscale
-                actual[i, j] = torch.exp(-val).item()
+                val = (torch.sin(math.pi / period * (a[i] - b[j])) / lengthscale) ** 2
+                actual[i, j] = torch.exp(-0.5 * val).item()
 
         res = kernel(a, b).evaluate()
         self.assertLess(torch.norm(res - actual), 1e-5)
+
+    def test_is_pd(self):
+        # ensures 1d input is positive definite with additional jitter
+        x = torch.randn(100).reshape(-1, 1)
+        kernel = PeriodicKernel()
+        with torch.no_grad():
+            K = kernel(x, x).evaluate() + 1e-4 * torch.eye(len(x))
+            eig = torch.symeig(K)[0]
+            self.assertTrue((eig > 0.0).all().item())
+
+    def test_multidimensional_inputs(self):
+        # test taken from issue #835
+        # ensures multidimensional input results in a positive definite kernel matrix, with additional jitter
+        x = torch.randn(1000, 2)
+        kernel = PeriodicKernel()
+        with torch.no_grad():
+            K = kernel(x, x).evaluate() + 1e-4 * torch.eye(len(x))
+            eig = torch.symeig(K)[0]
+            self.assertTrue((eig > 0.0).all().item())
 
     def test_batch(self):
         a = torch.tensor([[4, 2, 8], [1, 2, 3]], dtype=torch.float).view(2, 3, 1)
@@ -36,31 +55,25 @@ class TestPeriodicKernel(unittest.TestCase):
 
         actual = torch.zeros(2, 3, 2)
         for k in range(2):
-            for i in range(3):
-                for j in range(2):
-                    val = 2 * torch.pow(torch.sin(math.pi * (a[k, i] - b[k, j]) / period), 2) / lengthscale
-                    actual[k, i, j] = torch.exp(-val).item()
+            actual[k] = kernel(a[k], b[k]).evaluate()
 
         res = kernel(a, b).evaluate()
         self.assertLess(torch.norm(res - actual), 1e-5)
 
-    def test_batch_separate(self):
-        a = torch.tensor([[4, 2, 8], [1, 2, 3]], dtype=torch.float).view(2, 3, 1)
-        b = torch.tensor([[0, 2], [-1, 2]], dtype=torch.float).view(2, 2, 1)
-        period = torch.tensor([1, 2], dtype=torch.float).view(2, 1, 1)
-        lengthscale = torch.tensor([2, 1], dtype=torch.float).view(2, 1, 1)
-        kernel = PeriodicKernel(batch_shape=torch.Size([2])).initialize(lengthscale=lengthscale, period_length=period)
-        kernel.eval()
-
-        actual = torch.zeros(2, 3, 2)
-        for k in range(2):
-            for i in range(3):
-                for j in range(2):
-                    val = 2 * torch.pow(torch.sin(math.pi * (a[k, i] - b[k, j]) / period[k]), 2) / lengthscale[k]
-                    actual[k, i, j] = torch.exp(-val).item()
-
-        res = kernel(a, b).evaluate()
-        self.assertLess(torch.norm(res - actual), 1e-5)
+    # def test_batch_separate(self):
+    #     a = torch.tensor([[4, 2, 8], [1, 2, 3]], dtype=torch.float).view(2, 3, 1)
+    #     b = torch.tensor([[0, 2], [-1, 2]], dtype=torch.float).view(2, 2, 1)
+    #     period = torch.tensor([1, 2], dtype=torch.float).view(2, 1, 1)
+    #     lengthscale = torch.tensor([2, 1], dtype=torch.float).view(2, 1, 1)
+    #     kernel = PeriodicKernel(batch_shape=torch.Size([2])).initialize(lengthscale=lengthscale, period_length=period)
+    #     kernel.eval()
+    #
+    #     actual = torch.zeros(2, 3, 2)
+    #     for k in range(2):
+    #         actual[k] = kernel(a[k], b[k]).evaluate()
+    #
+    #     res = kernel(a, b).evaluate()
+    #     self.assertLess(torch.norm(res - actual), 1e-5)
 
 
 if __name__ == "__main__":

